### ISSUE TITLE
Unmount hooks should be called during the commit phase to ensure consistency with React

### ISSUE THREAD

**Issue by farwayer** [preactjs/preact#4299]
- [x] Check if updating to the latest Preact version resolves the issue

**Describe the bug**

I've run into issues several times that are caused by calling cleanup functions during the rendering phase. First time in the UI library (a similar case is described here https://github.com/preactjs/preact/issues/4222). The second time with a library that prohibits data modification at the rendering stage.

It would be nice for React compatibility and to prevent problems like this to call cleanup functions on hooks after rather than during rendering.

**Comment by farwayer** [comment]
Yes, you are right. I'm talking about calling the cleanup function of a child component while rendering the parent, if the child component is unmounted. The timing of the call is different from React and this sometimes causes problems.

```js
let Parent = () => {
  console.log('parent render start')
  useLayoutEffect(() => console.log('parent render end'))

  let [showChild, setShowChild] = useState(true)

  useEffect(() => {
    setTimeout(() => setShowChild(false), 1000)
  }, [])

  return showChild ? <Child/> : null
}

let Child = () => {
  useEffect(() => () => console.log('child cleanup'), [])
  return null
}
```

React:
```
parent render start
parent render end
parent render start
parent render end
child cleanup
```

Preact:
```
parent render start
parent render end
parent render start
child cleanup
parent render end
```

**Comment by farwayer** [comment]
Just checked v11-beta and the problem still exists. @JoviDeCroock You mentioned this issue in the PR for version 11. Was there a plan to make the child component cleanup sequence consistent with React?

**Comment by JoviDeCroock** [comment]
I had it set up as a candidate for v11 but never got to it, it's still an open issue worth exploring but I personally don't have bandwidth at this time.

**Comment by farwayer** [comment]
Ok, thanks! maybe I'll try to find time and see it myself
